File: old_code/src/utils/StepLexer.ts
/**
 * Unified StepLexer with zerocopy core implementation.
 * Consolidates standard, zerocopy, and optimized implementations into one.
 * Provides superior performance through memory arena allocation, string interning,
 * and object pooling while maintaining backward compatibility.
 */

// Memory infrastructure imports
import { MemoryArena } from '../memory/arena/MemoryArena';
import { StringInterner } from '../memory/strings/StringInterner';
import { ObjectPool, PoolableObject, ObjectFactory } from '../memory/pools/ObjectPool';

// Zerocopy token infrastructure
import { AlignedToken, AlignedTokenArray, TokenType, TokenSpan } from '../zerocopy/tokens/AlignedToken';

// Standard API imports (preserved for compatibility)
import { LexerPath } from './LexerPath';
import { Token } from './Token';
import { Terminal } from './Terminal';
import { TerminalMatch } from './TerminalMatch';
import { LexerOptions } from './LexerOptions';
import { IParserLexerSourceContainer, IParserLexerSourceLine } from './IParserLexerSource';
import { StepParser } from './StepParser';

/**
 * Pooled lexer path for zerocopy implementation
 */
class PooledLexerPath extends LexerPath implements PoolableObject {
  private _inUse: boolean = false;

  // Additional zerocopy state
  private _tokens: AlignedToken[] = [];
  private _score: number = 0;

  constructor() {
    super(); // Call parent LexerPath constructor
  }

  // PoolableObject interface methods
  reset(): void {
    this.setLexerPathId(LexerPath.NOTSET);
    this.setActiveLineNumber(0);
    this.setActiveCharacterNumber(0);
    this._tokens = [];
    this._score = 0;
    this._inUse = false;
  }

  isInUse(): boolean {
    return this._inUse;
  }

  setInUse(inUse: boolean): void {
    this._inUse = inUse;
  }

  // Zerocopy-specific methods
  getTokens(): AlignedToken[] {
    return this._tokens;
  }

  addToken(token: AlignedToken): void {
    this._tokens.push(token);
  }

  getScore(): number {
    return this._score;
  }

  setScore(score: number): void {
    this._score = score;
  }
}

/**
 * Factory for creating pooled lexer paths
 */
class LexerPathFactory implements ObjectFactory<PooledLexerPath> {
  create(): PooledLexerPath {
    return new PooledLexerPath();
  }

  reset(path: PooledLexerPath): void {
    path.reset();
  }

  validate(path: PooledLexerPath): boolean {
    return path.getLexerPathId() !== LexerPath.NOTSET;
  }
}

/**
 * Unified StepLexer class with zerocopy core.
 * Maintains the same API as the original StepLexer while internally using
 * zerocopy infrastructure for superior performance.
 */
export class StepLexer {
  // Zerocopy infrastructure (private, hidden from users)
  private arena: MemoryArena;
  private stringInterner: StringInterner;
  private pathPool: ObjectPool<PooledLexerPath>;
  private tokenArray: AlignedTokenArray;

  // Public API preserved for compatibility
  private sourceLinesContainer: IParserLexerSourceContainer;
  private stepParser: StepParser;
  private lexerOptions: LexerOptions;
  private lexerPathMap: Map<number, PooledLexerPath>;
  private invalidatedLexerPaths: PooledLexerPath[];
  private maximumLexerPathId: number;

  /**
   * Creates a new unified StepLexer instance.
   * Maintains the same constructor signature for backward compatibility.
   * @param stepParser The parser to use
   * @param lexerOptions The lexer options
   * @param sourceLinesContainer The source lines container
   */
  constructor(stepParser: StepParser, lexerOptions: LexerOptions, sourceLinesContainer: IParserLexerSourceContainer) {
    // Initialize zerocopy infrastructure
    this.initializeZeroCopyInfrastructure();

    // Preserve existing initialization logic
    this.stepParser = stepParser;
    this.lexerOptions = lexerOptions;
    this.sourceLinesContainer = sourceLinesContainer;
    this.lexerPathMap = new Map<number, PooledLexerPath>();
    this.invalidatedLexerPaths = [];
    this.maximumLexerPathId = LexerPath.NOTSET;
    this.reset();
  }

  /**
   * Initializes the zerocopy infrastructure.
   * Creates memory arena, string interner, and object pools.
   */
  private initializeZeroCopyInfrastructure(): void {
    // Create memory arena with 1MB initial size (proven beneficial)
    this.arena = new MemoryArena(1024 * 1024);

    // Create string interner for deduplication (proven beneficial)
    this.stringInterner = new StringInterner(this.arena);

    // Create object pool for lexer paths (proven beneficial)
    const pathFactory = new LexerPathFactory();
    this.pathPool = new ObjectPool(pathFactory, this.arena, 100, 1000);

    // Create aligned token array (proven beneficial)
    this.tokenArray = new AlignedTokenArray(this.arena, this.stringInterner, 1000);
  }

  /**
   * Resets the lexer.
   * Maintains the same behavior as the original implementation.
   */
  public reset(): void {
    // Clear existing paths and return them to pool
    for (const path of this.lexerPathMap.values()) {
      this.pathPool.release(path);
    }
    this.lexerPathMap.clear();

    // Clear invalidated paths and return them to pool
    for (const path of this.invalidatedLexerPaths) {
      this.pathPool.release(path);
    }
    this.invalidatedLexerPaths = [];

    this.maximumLexerPathId = LexerPath.NOTSET;
    this.newStartLexerPath();
  }

  /**
   * Gets the next tokens.
   * Maintains the same API as the original implementation.
   * @returns An iterator over token lists
   */
  public *nextTokens(): Generator<Token[]> {
    let tokenList: Token[] = [];
    let currentLexerPaths: PooledLexerPath[];

    while (this.lexerPathMap.size > 0) {
      tokenList = [];
      this.checkForParserLexerPathInvalidation(tokenList);
      this.checkForMergeLexerPaths(Array.from(this.lexerPathMap.values()), tokenList);
      currentLexerPaths = Array.from(this.lexerPathMap.values());

      for (const lexerPath of currentLexerPaths) {
        this.processNextToken(lexerPath, tokenList);
      }

      yield tokenList;
    }
  }

  /**
   * Invalidates a lexer path.
   * Maintains the same API as the original implementation.
   * @param lexerPathId The ID of the lexer path to invalidate
   */
  public invalidateLexerPath(lexerPathId: number): void {
    this.removeLexerPath(lexerPathId, true);
  }

  /**
   * Creates a new start lexer path using the object pool.
   */
  private newStartLexerPath(): void {
    const path = this.pathPool.acquire();
    this.maximumLexerPathId++;
    path.setLexerPathId(this.maximumLexerPathId);
    path.setActiveLineNumber(0);
    path.setActiveCharacterNumber(0);
    this.lexerPathMap.set(this.maximumLexerPathId, path);
  }

  /**
   * Removes a lexer path and optionally marks it as invalidated.
   * @param lexerPathId The ID of the path to remove
   * @param markAsInvalidated Whether to mark as invalidated
   */
  private removeLexerPath(lexerPathId: number, markAsInvalidated: boolean): void {
    const path = this.lexerPathMap.get(lexerPathId);
    if (path) {
      this.lexerPathMap.delete(lexerPathId);

      if (markAsInvalidated) {
        this.invalidatedLexerPaths.push(path);
      } else {
        this.pathPool.release(path);
      }
    }
  }

  /**
   * Checks for parser lexer path invalidation.
   * @param tokenList The token list to add invalidation tokens to
   */
  private checkForParserLexerPathInvalidation(tokenList: Token[]): void {
    if (this.lexerOptions.getReturnLexerPathTokens()) {
      for (const invalidLexerPath of this.invalidatedLexerPaths) {
        tokenList.push(new Token(
          invalidLexerPath.getLexerPathId(),
          new Terminal('LEXERPATH_REMOVED', ''),
          Token.LEXERPATH_EXTERN_REMOVED,
          invalidLexerPath.getActiveLineNumber(),
          invalidLexerPath.getActiveCharacterNumber(),
        ));
      }
    }

    // Return invalidated paths to pool
    for (const path of this.invalidatedLexerPaths) {
      this.pathPool.release(path);
    }
    this.invalidatedLexerPaths = [];
  }

  /**
   * Checks for merge opportunities between lexer paths.
   * Core functionality for handling ambiguous grammar efficiently.
   * @param lexerPaths The current lexer paths
   * @param tokenList The token list to add merge tokens to
   */
  private checkForMergeLexerPaths(lexerPaths: PooledLexerPath[], tokenList: Token[]): void {
    // Group paths by position to find merge candidates
    const pathsByPosition = new Map<string, PooledLexerPath[]>();

    for (const path of lexerPaths) {
      const positionKey = `${path.getActiveLineNumber()}-${path.getActiveCharacterNumber()}`;
      if (!pathsByPosition.has(positionKey)) {
        pathsByPosition.set(positionKey, []);
      }
      pathsByPosition.get(positionKey)!.push(path);
    }

    // Check for paths that can be merged
    for (const [_position, pathsAtPosition] of pathsByPosition.entries()) {
      if (pathsAtPosition.length > 1) {
        // Multiple paths at same position - potential merge opportunity
        const basePath = pathsAtPosition[0];

        for (let i = 1; i < pathsAtPosition.length; i++) {
          const mergePath = pathsAtPosition[i];

          // Check if paths have similar token sequences (simplified merge criteria)
          if (this.canMergePaths(basePath, mergePath)) {
            // Create merge token if lexer options require it
            if (this.lexerOptions.getReturnLexerPathTokens()) {
              tokenList.push(new Token(
                mergePath.getLexerPathId(),
                new Terminal('LEXERPATH_MERGE', ''),
                basePath.getLexerPathId().toString(),
                mergePath.getActiveLineNumber(),
                mergePath.getActiveCharacterNumber(),
              ));
            }

            // Remove the merged path
            this.removeLexerPath(mergePath.getLexerPathId(), false);
          }
        }
      }
    }
  }

  /**
   * Checks if two lexer paths can be merged based on their tokens and scores.
   * @param basePath The base path to merge into
   * @param mergePath The path to potentially merge
   * @returns True if paths can be merged
   */
  private canMergePaths(basePath: PooledLexerPath, mergePath: PooledLexerPath): boolean {
    // Simple merge criteria - paths at same position with similar scores
    const baseTokens = basePath.getTokens();
    const mergeTokens = mergePath.getTokens();

    // If token sequences are similar (basic comparison)
    if (baseTokens.length === mergeTokens.length) {
      const scoreDifference = Math.abs(basePath.getScore() - mergePath.getScore());
      return scoreDifference < 0.1; // Merge if scores are very close
    }

    return false;
  }

  /**
   * Processes the next token for a lexer path.
   * @param lexerPath The lexer path to process
   * @param tokenList The token list to add tokens to
   */
  private processNextToken(lexerPath: PooledLexerPath, tokenList: Token[]): void {
    // Get current source line
    const sourceLine = this.getCurrentSourceLine(lexerPath);
    if (!sourceLine) {
      this.removeLexerPath(lexerPath.getLexerPathId(), false);
      return;
    }

    // Get valid terminals from parser
    const validTerminals = this.stepParser.getValidTerminals(lexerPath);

    // Try to match terminals using zerocopy approach
    const matches = this.findTerminalMatches(sourceLine, lexerPath, validTerminals);

    if (matches.length === 0) {
      // No matches found, remove path
      this.removeLexerPath(lexerPath.getLexerPathId(), false);
    } else {
      // Process matches and create new paths if needed
      this.processTerminalMatches(lexerPath, matches, tokenList);
    }
  }

  /**
   * Gets the current source line for a lexer path.
   * @param lexerPath The lexer path
   * @returns The current source line or null
   */
  private getCurrentSourceLine(lexerPath: PooledLexerPath): IParserLexerSourceLine | null {
    const lineNumber = lexerPath.getActiveLineNumber();
    const lines = this.sourceLinesContainer.getSourceLines();
    return lines.find(line => line.getLineNumber() === lineNumber) || null;
  }

  /**
   * Finds terminal matches using zerocopy string operations.
   * @param sourceLine The source line to match against
   * @param lexerPath The current lexer path
   * @param validTerminals The valid terminals to try
   * @returns Array of terminal matches
   */
  private findTerminalMatches(
    sourceLine: IParserLexerSourceLine,
    lexerPath: PooledLexerPath,
    validTerminals: Terminal[],
  ): TerminalMatch[] {
    const matches: TerminalMatch[] = [];
    const lineContent = sourceLine.getContent();
    const startPos = lexerPath.getActiveCharacterNumber();

    for (const terminal of validTerminals) {
      const match = this.tryMatchTerminal(terminal, lineContent, startPos);
      if (match) {
        matches.push(match);
      }
    }

    return matches;
  }

  /**
   * Tries to match a terminal at the current position.
   * @param terminal The terminal to match
   * @param lineContent The interned line content
   * @param startPos The starting position
   * @returns A terminal match or null
   */
  private tryMatchTerminal(terminal: Terminal, lineContent: string, startPos: number): TerminalMatch | null {
    const input = lineContent.substring(startPos);
    const match = terminal.match(input);

    if (match && match.length > 0) {
      const value = match[1] || match[0];
      return new TerminalMatch(terminal, value, match[0].length);
    }

    return null;
  }

  /**
   * Processes terminal matches and creates tokens.
   * @param lexerPath The current lexer path
   * @param matches The terminal matches
   * @param tokenList The token list to add tokens to
   */
  private processTerminalMatches(
    lexerPath: PooledLexerPath,
    matches: TerminalMatch[],
    tokenList: Token[],
  ): void {
    if (matches.length === 0) {
      // No matches, create unknown token
      this.createUnknownToken(lexerPath, tokenList);
      return;
    }

    if (matches.length === 1) {
      // Single match, create token
      this.createTokenFromMatch(lexerPath, matches[0], tokenList);
      return;
    }

    // Multiple matches (ambiguity), handle by creating multiple paths
    this.handleMultipleMatches(lexerPath, matches, tokenList);
  }

  /**
   * Creates an unknown token using zerocopy approach.
   * @param lexerPath The lexer path
   * @param tokenList The token list to add the token to
   */
  private createUnknownToken(lexerPath: PooledLexerPath, tokenList: Token[]): void {
    const sourceLine = this.getCurrentSourceLine(lexerPath);
    if (!sourceLine) {
      this.removeLexerPath(lexerPath.getLexerPathId(), false);
      return;
    }

    const lineContent = sourceLine.getContent();
    const charPos = lexerPath.getActiveCharacterNumber();

    if (charPos >= lineContent.length) {
      // End of line, move to next line
      lexerPath.setActiveLineNumber(lexerPath.getActiveLineNumber() + 1);
      lexerPath.setActiveCharacterNumber(0);
      return;
    }

    const value = lineContent.charAt(charPos);

    const unknownToken = new Token(
      lexerPath.getLexerPathId(),
      new Terminal('UNKNOWN', ''),
      value, // Use the actual string value, not interned ID
      lexerPath.getActiveLineNumber(),
      lexerPath.getActiveCharacterNumber(),
    );

    lexerPath.setActiveCharacterNumber(charPos + 1);
    tokenList.push(unknownToken);
  }

  /**
   * Creates a token from a terminal match using zerocopy approach.
   * @param lexerPath The lexer path
   * @param match The terminal match
   * @param tokenList The token list to add the token to
   */
  private createTokenFromMatch(lexerPath: PooledLexerPath, match: TerminalMatch, tokenList: Token[]): void {
    const value = match.getValue(); // Use the actual string value, not the interned ID

    const token = new Token(
      lexerPath.getLexerPathId(),
      match.getTerminal(),
      value,
      lexerPath.getActiveLineNumber(),
      lexerPath.getActiveCharacterNumber(),
    );

    lexerPath.setActiveCharacterNumber(lexerPath.getActiveCharacterNumber() + match.getLength());
    tokenList.push(token);

    // Create aligned token for zerocopy tracking
    const span: TokenSpan = {
      start: {
        line: lexerPath.getActiveLineNumber(),
        column: lexerPath.getActiveCharacterNumber(),
        offset: lexerPath.getActiveCharacterNumber(),
      },
      end: {
        line: lexerPath.getActiveLineNumber(),
        column: lexerPath.getActiveCharacterNumber() + match.getLength(),
        offset: lexerPath.getActiveCharacterNumber() + match.getLength(),
      },
    };

    const alignedToken = this.tokenArray.push(
      TokenType.IDENTIFIER, // Default type, could be enhanced based on terminal
      value,
      span,
    );
    lexerPath.addToken(alignedToken);
  }

  /**
   * Handles multiple terminal matches (ambiguity) using object pooling.
   * @param lexerPath The lexer path
   * @param matches The terminal matches
   * @param tokenList The token list to add tokens to
   */
  private handleMultipleMatches(lexerPath: PooledLexerPath, matches: TerminalMatch[], tokenList: Token[]): void {
    // Use the first match for the current path
    this.createTokenFromMatch(lexerPath, matches[0], tokenList);

    // Create new paths for the other matches using object pool
    for (let i = 1; i < matches.length; i++) {
      const newPath = this.pathPool.acquire();
      this.maximumLexerPathId++;

      // Copy state from current path
      newPath.setLexerPathId(this.maximumLexerPathId);
      newPath.setActiveLineNumber(lexerPath.getActiveLineNumber());
      newPath.setActiveCharacterNumber(lexerPath.getActiveCharacterNumber());

      // Add to path map
      this.lexerPathMap.set(this.maximumLexerPathId, newPath);

      // Create token for this path
      this.createTokenFromMatch(newPath, matches[i], tokenList);
    }
  }
}

File: src/StepParser/Lexer/StepLexer.cs
using StepParser.Core;
using StepParser.Lexer;
using StepParser.Memory;

namespace StepParser.Lexer;

/// <summary>
/// Factory for creating pooled lexer paths.
/// </summary>
public class LexerPathFactory : IObjectFactory<LexerPath>
{
    /// <inheritdoc/>
    public LexerPath Create()
    {
        return new LexerPath();
    }

    /// <inheritdoc/>
    public void Reset(LexerPath obj)
    {
        obj.Reset();
    }

    /// <inheritdoc/>
    public bool Validate(LexerPath obj)
    {
        return obj.LexerPathId != LexerPath.NOTSET;
    }
}

/// <summary>
/// Unified StepLexer class with zero-copy core.
/// Maintains the same API as the original StepLexer while internally using
/// zero-copy infrastructure for superior performance.
/// </summary>
public class StepLexer
{
    // Zero-copy infrastructure (private, hidden from users)
    private readonly MemoryArena _arena;
    private readonly StringInterner _stringInterner;
    private readonly ObjectPool<LexerPath> _pathPool;

    // Public API preserved for compatibility
    private readonly IParserLexerSourceContainer _sourceLinesContainer;
    private readonly Core.StepParser _stepParser;
    private readonly LexerOptions _lexerOptions;
    private readonly Dictionary<int, LexerPath> _lexerPathMap;
    private readonly List<LexerPath> _invalidatedLexerPaths;
    private int _maximumLexerPathId;

    /// <summary>
    /// Initializes a new instance of the <see cref="StepLexer"/> class.
    /// </summary>
    /// <param name="stepParser">The parser to use.</param>
    /// <param name="lexerOptions">The lexer options.</param>
    /// <param name="sourceLinesContainer">The source lines container.</param>
    public StepLexer(Core.StepParser stepParser, LexerOptions lexerOptions, IParserLexerSourceContainer sourceLinesContainer)
    {
        // Initialize zero-copy infrastructure
        _arena = new MemoryArena(1024 * 1024); // 1MB initial size
        _stringInterner = new StringInterner(_arena);

        // Create object pool for lexer paths
        var pathFactory = new LexerPathFactory();
        _pathPool = new ObjectPool<LexerPath>(pathFactory, _arena, 100, 1000);

        // Preserve existing initialization logic
        _stepParser = stepParser;
        _lexerOptions = lexerOptions;
        _sourceLinesContainer = sourceLinesContainer;
        _lexerPathMap = new Dictionary<int, LexerPath>();
        _invalidatedLexerPaths = new List<LexerPath>();
        _maximumLexerPathId = LexerPath.NOTSET;
        Reset();
    }

    /// <summary>
    /// Resets the lexer.
    /// </summary>
    public void Reset()
    {
        // Clear existing paths and return them to pool
        foreach (var path in _lexerPathMap.Values)
        {
            _pathPool.Release(path);
        }
        _lexerPathMap.Clear();

        // Clear invalidated paths and return them to pool
        foreach (var path in _invalidatedLexerPaths)
        {
            _pathPool.Release(path);
        }
        _invalidatedLexerPaths.Clear();

        _maximumLexerPathId = LexerPath.NOTSET;
        NewStartLexerPath();
    }

    /// <summary>
    /// Gets the next tokens.
    /// </summary>
    /// <returns>An enumerable of token lists.</returns>
    public IEnumerable<List<Token>> NextTokens()
    {
        var tokenList = new List<Token>();
        var currentLexerPaths = new List<LexerPath>();

        while (_lexerPathMap.Count > 0)
        {
            tokenList.Clear();
            CheckForParserLexerPathInvalidation(tokenList);
            CheckForMergeLexerPaths(_lexerPathMap.Values.ToList(), tokenList);
            currentLexerPaths.Clear();
            currentLexerPaths.AddRange(_lexerPathMap.Values);

            foreach (var lexerPath in currentLexerPaths)
            {
                ProcessNextToken(lexerPath, tokenList);
            }

            yield return new List<Token>(tokenList);
        }
    }

    /// <summary>
    /// Invalidates a lexer path.
    /// </summary>
    /// <param name="lexerPathId">The ID of the lexer path to invalidate.</param>
    public void InvalidateLexerPath(int lexerPathId)
    {
        RemoveLexerPath(lexerPathId, true);
    }

    /// <summary>
    /// Creates a new start lexer path using the object pool.
    /// </summary>
    private void NewStartLexerPath()
    {
        var path = _pathPool.Acquire();
        _maximumLexerPathId++;
        path.LexerPathId = _maximumLexerPathId;
        path.ActiveLineNumber = 0;
        path.ActiveCharacterNumber = 0;
        _lexerPathMap[_maximumLexerPathId] = path;
    }

    /// <summary>
    /// Removes a lexer path and optionally marks it as invalidated.
    /// </summary>
    /// <param name="lexerPathId">The ID of the path to remove.</param>
    /// <param name="markAsInvalidated">Whether to mark as invalidated.</param>
    private void RemoveLexerPath(int lexerPathId, bool markAsInvalidated)
    {
        if (_lexerPathMap.TryGetValue(lexerPathId, out var path))
        {
            _lexerPathMap.Remove(lexerPathId);

            if (markAsInvalidated)
            {
                _invalidatedLexerPaths.Add(path);
            }
            else
            {
                _pathPool.Release(path);
            }
        }
    }

    /// <summary>
    /// Checks for parser lexer path invalidation.
    /// </summary>
    /// <param name="tokenList">The token list to add invalidation tokens to.</param>
    private void CheckForParserLexerPathInvalidation(List<Token> tokenList)
    {
        if (_lexerOptions.ReturnLexerPathTokens)
        {
            foreach (var invalidLexerPath in _invalidatedLexerPaths)
            {
                tokenList.Add(new Token(
                    invalidLexerPath.LexerPathId,
                    new Terminal("LEXERPATH_REMOVED", ""),
                    Token.LEXERPATH_EXTERN_REMOVED,
                    invalidLexerPath.ActiveLineNumber,
                    invalidLexerPath.ActiveCharacterNumber));
            }
        }

        // Return invalidated paths to pool
        foreach (var path in _invalidatedLexerPaths)
        {
            _pathPool.Release(path);
        }
        _invalidatedLexerPaths.Clear();
    }

    /// <summary>
    /// Checks for merge opportunities between lexer paths.
    /// </summary>
    /// <param name="lexerPaths">The current lexer paths.</param>
    /// <param name="tokenList">The token list to add merge tokens to.</param>
    private void CheckForMergeLexerPaths(List<LexerPath> lexerPaths, List<Token> tokenList)
    {
        // Group paths by position to find merge candidates
        var pathsByPosition = new Dictionary<string, List<LexerPath>>();

        foreach (var path in lexerPaths)
        {
            var positionKey = $"{path.ActiveLineNumber}-{path.ActiveCharacterNumber}";
            if (!pathsByPosition.ContainsKey(positionKey))
            {
                pathsByPosition[positionKey] = new List<LexerPath>();
            }
            pathsByPosition[positionKey].Add(path);
        }

        // Check for paths that can be merged
        foreach (var pathsAtPosition in pathsByPosition.Values)
        {
            if (pathsAtPosition.Count > 1)
            {
                // Multiple paths at same position - potential merge opportunity
                var basePath = pathsAtPosition[0];

                for (int i = 1; i < pathsAtPosition.Count; i++)
                {
                    var mergePath = pathsAtPosition[i];

                    // Check if paths can be merged (simplified criteria)
                    if (CanMergePaths(basePath, mergePath))
                    {
                        // Create merge token if lexer options require it
                        if (_lexerOptions.ReturnLexerPathTokens)
                        {
                            tokenList.Add(new Token(
                                mergePath.LexerPathId,
                                new Terminal("LEXERPATH_MERGE", ""),
                                basePath.LexerPathId.ToString(),
                                mergePath.ActiveLineNumber,
                                mergePath.ActiveCharacterNumber));
                        }

                        // Remove the merged path
                        RemoveLexerPath(mergePath.LexerPathId, false);
                    }
                }
            }
        }
    }

    /// <summary>
    /// Checks if two lexer paths can be merged.
    /// </summary>
    /// <param name="basePath">The base path to merge into.</param>
    /// <param name="mergePath">The path to potentially merge.</param>
    /// <returns>True if paths can be merged.</returns>
    private bool CanMergePaths(LexerPath basePath, LexerPath mergePath)
    {
        // Simple merge criteria - paths at same position
        return basePath.ActiveLineNumber == mergePath.ActiveLineNumber &&
               basePath.ActiveCharacterNumber == mergePath.ActiveCharacterNumber;
    }

    /// <summary>
    /// Processes the next token for a lexer path.
    /// </summary>
    /// <param name="lexerPath">The lexer path to process.</param>
    /// <param name="tokenList">The token list to add tokens to.</param>
    private void ProcessNextToken(LexerPath lexerPath, List<Token> tokenList)
    {
        // Get current source line
        var sourceLine = GetCurrentSourceLine(lexerPath);
        if (sourceLine == null)
        {
            RemoveLexerPath(lexerPath.LexerPathId, false);
            return;
        }

        // Get valid terminals from parser
        var validTerminals = _stepParser.GetValidTerminals(lexerPath);

        // Try to match terminals
        var matches = FindTerminalMatches(sourceLine, lexerPath, validTerminals);

        if (matches.Count == 0)
        {
            // No matches found, create unknown token or remove path
            CreateUnknownToken(lexerPath, tokenList);
        }
        else
        {
            // Process matches and create new paths if needed
            ProcessTerminalMatches(lexerPath, matches, tokenList);
        }
    }

    /// <summary>
    /// Gets the current source line for a lexer path.
    /// </summary>
    /// <param name="lexerPath">The lexer path.</param>
    /// <returns>The current source line or null.</returns>
    private IParserLexerSourceLine? GetCurrentSourceLine(LexerPath lexerPath)
    {
        return _sourceLinesContainer.SourceLines
            .FirstOrDefault(line => line.LineNumber == lexerPath.ActiveLineNumber);
    }

    /// <summary>
    /// Finds terminal matches using string operations.
    /// </summary>
    /// <param name="sourceLine">The source line to match against.</param>
    /// <param name="lexerPath">The current lexer path.</param>
    /// <param name="validTerminals">The valid terminals to try.</param>
    /// <returns>Array of terminal matches.</returns>
    private List<TerminalMatch> FindTerminalMatches(
        IParserLexerSourceLine sourceLine,
        LexerPath lexerPath,
        IEnumerable<Terminal> validTerminals)
    {
        var matches = new List<TerminalMatch>();
        var lineContent = sourceLine.Content;
        var startPos = lexerPath.ActiveCharacterNumber;

        foreach (var terminal in validTerminals)
        {
            var match = TryMatchTerminal(terminal, lineContent, startPos);
            if (match != null)
            {
                matches.Add(match);
            }
        }

        return matches;
    }

    /// <summary>
    /// Tries to match a terminal at the current position.
    /// </summary>
    /// <param name="terminal">The terminal to match.</param>
    /// <param name="lineContent">The line content.</param>
    /// <param name="startPos">The starting position.</param>
    /// <returns>A terminal match or null.</returns>
    private TerminalMatch? TryMatchTerminal(Terminal terminal, string lineContent, int startPos)
    {
        if (startPos >= lineContent.Length)
            return null;

        var input = lineContent.Substring(startPos);
        var match = terminal.MatchInput(input);

        if (match != null && match.Success)
        {
            var value = match.Groups.Count > 1 ? match.Groups[1].Value : match.Groups[0].Value;
            return new TerminalMatch(terminal, value, match.Length);
        }

        return null;
    }

    /// <summary>
    /// Creates an unknown token.
    /// </summary>
    /// <param name="lexerPath">The lexer path.</param>
    /// <param name="tokenList">The token list to add the token to.</param>
    private void CreateUnknownToken(LexerPath lexerPath, List<Token> tokenList)
    {
        var sourceLine = GetCurrentSourceLine(lexerPath);
        if (sourceLine == null)
        {
            RemoveLexerPath(lexerPath.LexerPathId, false);
            return;
        }

        var lineContent = sourceLine.Content;
        var charPos = lexerPath.ActiveCharacterNumber;

        if (charPos >= lineContent.Length)
        {
            // End of line, move to next line
            lexerPath.ActiveLineNumber++;
            lexerPath.ActiveCharacterNumber = 0;
            return;
        }

        var value = lineContent.Substring(charPos, 1);

        var unknownToken = new Token(
            lexerPath.LexerPathId,
            new Terminal("UNKNOWN", ""),
            value,
            lexerPath.ActiveLineNumber,
            lexerPath.ActiveCharacterNumber);

        lexerPath.ActiveCharacterNumber++;
        tokenList.Add(unknownToken);
    }

    /// <summary>
    /// Processes terminal matches and creates tokens.
    /// </summary>
    /// <param name="lexerPath">The current lexer path.</param>
    /// <param name="matches">The terminal matches.</param>
    /// <param name="tokenList">The token list to add tokens to.</param>
    private void ProcessTerminalMatches(
        LexerPath lexerPath,
        List<TerminalMatch> matches,
        List<Token> tokenList)
    {
        if (matches.Count == 1)
        {
            // Single match, create token
            CreateTokenFromMatch(lexerPath, matches[0], tokenList);
        }
        else
        {
            // Multiple matches (ambiguity), handle by creating multiple paths
            HandleMultipleMatches(lexerPath, matches, tokenList);
        }
    }

    /// <summary>
    /// Creates a token from a terminal match.
    /// </summary>
    /// <param name="lexerPath">The lexer path.</param>
    /// <param name="match">The terminal match.</param>
    /// <param name="tokenList">The token list to add the token to.</param>
    private void CreateTokenFromMatch(LexerPath lexerPath, TerminalMatch match, List<Token> tokenList)
    {
        var value = _stringInterner.Intern(match.Value);

        var token = new Token(
            lexerPath.LexerPathId,
            match.Terminal,
            value,
            lexerPath.ActiveLineNumber,
            lexerPath.ActiveCharacterNumber);

        lexerPath.ActiveCharacterNumber += match.Length;
        tokenList.Add(token);
    }

    /// <summary>
    /// Handles multiple terminal matches (ambiguity) using object pooling.
    /// </summary>
    /// <param name="lexerPath">The lexer path.</param>
    /// <param name="matches">The terminal matches.</param>
    /// <param name="tokenList">The token list to add tokens to.</param>
    private void HandleMultipleMatches(LexerPath lexerPath, List<TerminalMatch> matches, List<Token> tokenList)
    {
        // Use the first match for the current path
        CreateTokenFromMatch(lexerPath, matches[0], tokenList);

        // Create new paths for the other matches using object pool
        for (int i = 1; i < matches.Count; i++)
        {
            var newPath = _pathPool.Acquire();
            _maximumLexerPathId++;

            // Copy state from current path
            newPath.LexerPathId = _maximumLexerPathId;
            newPath.ActiveLineNumber = lexerPath.ActiveLineNumber;
            newPath.ActiveCharacterNumber = lexerPath.ActiveCharacterNumber;

            // Add to path map
            _lexerPathMap[_maximumLexerPathId] = newPath;

            // Create token for this path
            CreateTokenFromMatch(newPath, matches[i], tokenList);
        }
    }

    /// <summary>
    /// Releases all resources used by the lexer.
    /// </summary>
    public void Dispose()
    {
        _pathPool?.Dispose();
        _arena?.Dispose();
    }
}